m4opt/models/math.py

```diff
@@ -0,0 +1,247 @@
+    (-1.5,1.5)
+
+    This also works with models that have dimensionful input:
+
+    >>> _get_intervals(models.Box1D(width=3*u.m))
+    (<Quantity -1.5 m>,<Quantity 1.5 m>)
+
+    For unbounded models, `_get_intervals` returns an infinite interval:
+
+    >>> _get_intervals(models.Exponential1D())
+    (-inf,inf)
+    >>> _get_intervals(models.Exponential1D(tau=1*u.m))
+    (<Quantity -inf m>,<Quantity inf m>)
+
+    Compound models are supported too. Supported operations include function
+    composition (`|`):
+
+    >>> _get_intervals(models.Box1D(width=3) | models.Shift(offset=2))
+    (0.5,3.5)
+    >>> _get_intervals(models.Box1D(width=3*u.m) | models.Shift(offset=2*u.m))
+    (<Quantity 0.5 m>,<Quantity 3.5 m>)
+
+    Multiplication (`*`) results in an intersection:
+
+    >>> model1 = models.Box1D(x_0=-0.5, width=3)
+    >>> model2 = models.Box1D(x_0=0.5, width=3)
+    >>> _get_intervals(model1)
+    (-2.0,1.0)
+    >>> _get_intervals(model2)
+    (-1.0,2.0)
+    >>> _get_intervals(model1 * model2)
+    (-1.0,1.0)
+    >>> model1 = models.Box1D(x_0=-0.5*u.m, width=3*u.m)
+    >>> model2 = models.Box1D(x_0=0.5*u.m, width=3*u.m)
+    >>> _get_intervals(model1 * model2)
+    (<Quantity -1. m>,<Quantity 1. m>)
+
+    Addition (`+`) and subtraction (`-`) result in a union, but holes are kept.
+
+    >>> model1 = models.Box1D(x_0=-0.5, width=3)
+    >>> model2 = models.Box1D(x_0=0.5, width=3)
+    >>> _get_intervals(model1 + model2)
+    (-2.0,-1.0) | (-1.0,2.0)
+    >>> _get_intervals(model1 - model2)
+    (-2.0,-1.0) | (-1.0,2.0)
+    >>> model1 = models.Box1D(x_0=-0.5*u.m, width=3*u.m)
+    >>> model2 = models.Box1D(x_0=0.5*u.m, width=3*u.m)
+    >>> _get_intervals(model1 + model2)
+    (<Quantity -2. m>,<Quantity -1. m>) | (<Quantity -1. m>,<Quantity 2. m>)
+    >>> _get_intervals(model1 - model2)
+    (<Quantity -2. m>,<Quantity -1. m>) | (<Quantity -1. m>,<Quantity 2. m>)
+    """
+    if isinstance(model, CompoundModel):
+        if model.op == '|':
+            return _get_intervals(model.left).apply(_map_interval(model.right))
+        elif model.op == '*':
+            return _get_intervals(model.left) & _get_intervals(model.right)
+        elif model.op in {'+', '-', '/'}:
+            lhs = portion.IntervalDict({_get_intervals(model.left): 'lhs'})
+            rhs = portion.IntervalDict({_get_intervals(model.right): 'rhs'})
+            union = lhs | rhs
+            return portion.Interval(
+                *(intervals.apply(_to_open_interval) for intervals in union))
+        else:
+            raise NotImplementedError(f'operation {model.op} not supported')
+    else:
+        try:
+            bbox = model.bounding_box
+        except NotImplementedError:
+            unit = _get_1d_units_for_dict(model.input_units)
+            return portion.open(-np.inf * unit, np.inf * unit)
+        else:
+            (lo, hi), = bbox
+            return portion.open(lo, hi)
+
+
+def integrate(model: Model, *,
```

Comment:
this only works on  models that have a single `x` dimension as written, it seems to me.  Maybe this should be:
```suggestion
def integrate1d(model: Model, *,
```
or similar? I know this is the majority of model cases, but it's still important to clarify that it doesn't work for e.g. the 2D image models

Reply:
Done.